### DISCUSSION TITLE
Welcome to falco Discussions!

### DISCUSSION THREAD

**Discussion by Tobi-De** [Tobi-De/falco#9] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by hosfe** [comment]
Hello,

I am currently testing the tool Falco/CRUD for my model. I find the approach very helpful and can imagine using it in my projects. One point I don’t understand, however, is why the HTML files are overwritten. When I generate a model with the command "install-crud-utils", the files views.py and urls.py are amended accordingly. However, the HTML files are overwritten for each model. This is also documented: "Unlike the Python code, the generated HTML templates will overwrite any existing ones."

From my perspective, it would be helpful to have a specific output directory for the HTML files. This way, the views and urls would not need to be edited again.

Why is the current solution set up this way? What might I have misunderstood?

Thanks for the feedback.
Felix

**Reply by Tobi-De** [reply]
Hi Felix, thanks for your interest in Falco.

> When I generate a model with the command "install-crud-utils", the files views.py and urls.py are amended accordingly.

I'm guessing you mean `falco crud`? The `install-crud-utils` command should not touch any HTML files.

I didn't consider the idea of adding an option for an output directory. I'm not sure how this would work. Would the output only be related to the HTML code? Wouldn't that be confusing? 🤔

Now, regarding the statement `"Unlike the Python code, the generated HTML templates will overwrite any existing ones."`

Here are my reasons:

1. It's less troublesome to update Python than HTML. Even if I append new Python code to the end of a file, it likely won't break the app. With HTML, it will surely break the layouts (and be hard to navigate), especially if a lot of HTML code is generated (for example, if you have custom CRUD templates).
2. The main point of the `--only-html` option is in case you are past the prototyping phase and have [written custom CRUD templates](https://falco.oluwatobi.dev/the_cli/crud.html#custom-templates) for your app. Now, you want to overwrite the old ones. That's the main use case I had in mind when adding this CLI option. I would assume that at that point, you would want to override your existing HTML code with the new code.

Maybe I'm misunderstanding what you are saying, but if I'm generating the CRUD code a second time for the same models, why would I want the HTML in a different folder? That would mean having two pairs of HTML for the same Python code. Or maybe you have existing HTML files but only want the Python code, in which case there is the `--only-python` option.
I'm sorry, but I'm not really understanding the use case here. Can you please provide more details on what you are trying to achieve?

Nevertheless, I'd be more than willing to change this behavior if it's really annoying. Thank you :)

**Comment by hosfe** [comment]
Hello Tobi,

Thank you for your prompt response, and I apologize for not being clear earlier. I am not a software developer, so perhaps I misunderstood something in the guide!

Here's an attempt to explain my idea better:

**0. Project Setup**
I used the Falco CLI to set up the project:
- `falco start-project myproject`
- `falco start-app card2brain`

**1. models.py**
Three models were created:
```python
class Word(TimeStampedModel):
class Sentence(TimeStampedModel):
class WordType(TimeStampedModel):
```

**2. CRUD for Your Model**
For all three models, I executed the corresponding command:
- `falco crud card2brain.word --entry-point --skip-git-check`
- `falco crud card2brain.sentence --entry-point --skip-git-check`
- `falco crud card2brain.wordtype --entry-point --skip-git-check`

**3. HTML Files**
The necessary views.py and urls.py were successfully created and attached. During this process, a set of HTML files was created in the template directory:
- `create.html, delete.html, index.html, update.html`
These were created last for the `model.wordtype`. The previously created HTML files for `model.word` and `model.sentence` were overwritten.

**My Concern and Thoughts:**
Manually creating HTML pages is labor-intensive and error-prone. It would be helpful if I didn't have to recreate them each time, especially when deploying CRUD HTML files which are almost identical and contain the same HTML code.

**Current Understanding:**
**_"Unlike the Python code, the generated HTML templates will overwrite any existing ones."_**

**Assumption:**
The current implementation of "falco crud" does not allow creating all necessary files, including HTML files, in one step.

**Current Solution:**
I am currently using the "falco crud" command to save the created HTML files into a subdirectory:
- `templates/card2brain/word`
- `templates/card2brain/sentence`
- `templates/card2brain/wordtype`

Afterwards, I manually adjust the views.py and urls.py so that the data can be accessed correctly.

**Example in views.py:**
```python
def word_detail(request: HttpRequest, pk: int):
    word = get_object_or_404(Word.objects, pk=pk)
    return TemplateResponse(
        request,
        "card2brain/word/detail.html",
        context={"word": word},
    )
```

**urls.py:**
```python
urlpatterns = [
    path("word", views.word_index, name="word_index"),
    path("word_create/", views.word_create, name="word_create"),
    path("<int:pk>/word_detail/", views.word_detail, name="word_detail"),
    path("<int:pk>/word_update/", views.word_update, name="word_update"),
    path("<int:pk>/word_delete/", views.word_delete, name="word_delete"),
]
```

Perhaps I have misunderstood something, as I am not a professional in this field. It would be helpful for me to be able to create the HTML files seamlessly in one go.

Thank you for your support,

Felix

**Reply by Tobi-De** [reply]
> I am not a software developer, so perhaps I misunderstood something in the guide!

I understand, it can be confusing for beginners and I wasn't expecting non-developers to use it :smile: 

> For all three models, I executed the corresponding command:
> 
>     * `falco crud card2brain.word --entry-point --skip-git-check`
> 
>     * `falco crud card2brain.sentence --entry-point --skip-git-check`
> 
>     * `falco crud card2brain.wordtype --entry-point --skip-git-check`


I think I see the issue now. Your HTML files were overridden because you used the `--entry-point` option for all three commands. This likely resulted in incorrect views and URLs being generated. The `--entry-point` option is meant to be used for only one model per Django app, not all models. Using it for all models will make all three models have the same URLs, Python views, and HTML files. You may not notice the issue with the Python part until you run the app, but since the HTML files are overridden, you will only see a single pair of HTML files, and they will be for the last model.

I suggest running something like this for each model:

```sh
falco crud card2brain.word --skip-git-check
```

This will make things simpler, and everything should be generated successfully. If these three models are all the models in your app, you can just run:

```sh
falco crud card2brain
```

This will do the work for all models in the app. Assuming you haven't made too many changes, delete the `views.py`, `urls.py`, `admin.py`, and the HTML files, and then run:

```sh
falco crud card2brain --skip-git-check
```

to regenerate everything for the three models.

I hope this helps! Let me know if you need further clarification.

**Comment by hosfe** [comment]
Hello Tobi,
many thanks for your support! I will test this out in the coming days and then give you brief feedback. I am confident it will solve the problem.
I hope I'm not annoying you too much with my questions.
Best regards,
Felix

**Reply by Tobi-De** [reply]
No worries Felix, happy to help :)
